library/build/tools.pl

```diff
@@ -431,15 +438,30 @@
     getenv('CC', CC),
     !.
 default_c_compiler(CC) :-
-    preferred_c_compiler(CC),
-    has_program(path(CC), _),
+    preferred_c_compiler(CC0),
+    has_program(CC0, CC),
+    !.
+
+default_cxx_compiler(CXX) :-
+    getenv('CXX', CXX),
+    !.
+default_cxx_compiler(CXX) :-
+    preferred_cxx_compiler(CXX0),
+    has_program(CXX0, CXX),
     !.
 
 preferred_c_compiler(CC) :-
     current_prolog_flag(c_cc, CC).
-preferred_c_compiler(gcc).
-preferred_c_compiler(clang).
-preferred_c_compiler(cc).
+preferred_c_compiler(path(gcc)).
```

Comment:
I could find no good reason for the path() wrapper.   has_program/2 adds this if necessary.   If this fixes some issue I think it is has_program/2 that needs fixing.    So, I skipped this.

Reply:
Line 521 (program_install_prefix/1) has `has_program(path(Name), EXE),` ... is it wrong?

`has_program/2` uses `pack_build_path(...)`, not `path(...)` ... presumably the former is better?